### PR TITLE
Video Pipeline hardware decoding support

### PR DESCRIPTION
**Pull request by Indumathi31**

### CODE REVIEW

**Review by shobana-mcw — COMMENTED** [review]
We need to handle the Error messages better. 
Will review the files once more and update comments if any.

File: rocAL/rocAL/source/hardware_video_decoder.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    return select_frame_pts;
+}
+
+int HardWareVideoDecoder::hw_decoder_init(AVCodecContext *ctx, const enum AVHWDeviceType type, AVBufferRef *hw_device_ctx)
+{
+    int err = 0;
+    if ((err = av_hwdevice_ctx_create(&hw_device_ctx, type, NULL, NULL, 0)) < 0)
+        return err;
+    ctx->hw_device_ctx = av_buffer_ref(hw_device_ctx);
+    return err;
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/hardware_video_decoder.cpp:49]: Handle this if condition better. We can have just one return statement.

File: rocAL/rocAL/source/hardware_video_decoder.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+static enum AVPixelFormat hwPixelFormat;
+enum AVPixelFormat get_hw_format(AVCodecContext *ctx, const enum AVPixelFormat *pix_fmts)
+{
+    const enum AVPixelFormat *p;
+    for (p = pix_fmts; *p != -1; p++) {
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/hardware_video_decoder.cpp:56]: Why are we iterating in const enum da? Did you not get any error while running this? 
Is this function called?

File: rocAL/rocAL/include/video_decoder.h [hunk 3]
```diff
@@ -36,13 +36,18 @@ extern "C"
...
+#include <libavutil/avassert.h>
+#include <libavutil/imgutils.h>
 }
 #endif
 #include "parameter_factory.h"
 
 enum class VideoDecoderType
 {
-    FFMPEG_VIDEO = 0, //!< Can decode video stream
+    FFMPEG_SOFTWARE_DECODE = 0, //!< Can decode video stream
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/include/video_decoder.h:49]: Done

File: rocAL/rocAL/source/hardware_video_decoder.cpp [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+    // Initialize the SwsContext 
+    SwsContext *swsctx = nullptr;
+    if ((out_width != _codec_width) || (out_height != _codec_height) || (out_pix_format != _dec_pix_fmt))
+    {
+        swsctx = sws_getCachedContext(nullptr, _codec_width, _codec_height, _dec_pix_fmt,
+                                      out_width, out_height, out_pix_format, SWS_BILINEAR, nullptr, nullptr, nullptr);
+        if (!swsctx)
+        {
+            ERR("Fail to get sws_getCachedContext");
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/source/hardware_video_decoder.cpp:77]: Done & replicated the same for all error messages

File: rocAL/rocAL/source/video_label_reader.cpp [hunk 5]
```diff
@@ -57,10 +57,9 @@ bool VideoLabelReader::exists(const std::string &frame_name)
...
 void VideoLabelReader::add(std::string frame_name, int label, unsigned int video_frame_count, unsigned int start_frame)
 {
-    std::vector<unsigned> props;
-    open_video_context(frame_name.c_str(), props);
-    unsigned frame_count = video_frame_count ? video_frame_count : props[2];
-    if ((video_frame_count + start_frame) > props[2])
+    open_video_context(frame_name.c_str(), _props);
+    unsigned frame_count = video_frame_count ? video_frame_count : _props[2];
+    if ((video_frame_count + start_frame) > _props[2])
         THROW("The given frame numbers in txt file exceeds the maximum frames in the video" + frame_name)
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply rocAL/rocAL/source/video_label_reader.cpp:63]: Its error only.

File: utilities/rali/rali_video_unittests/rali_video_unittests.cpp [hunk 6]
```diff
@@ -84,10 +84,8 @@ int main(int argc, const char **argv)
     bool enable_timestamps = true;
     bool enable_sequence_rearrange = false;
     bool is_output = true;
-    unsigned video_mode = 0;
-    // auto decoder_mode = ((video_mode == 1) ? RaliDecodeDevice::RALI_HW_DECODE : RaliDecodeDevice::RALI_SW_DECODE); // Hardware decoder support will be added in future
-    auto decoder_mode = RaliDecodeDevice::RALI_SW_DECODE;
-
+    unsigned hardware_decode_mode = 0;
+    auto decoder_mode = ((hardware_decode_mode == 1) ? RaliDecodeDevice::RALI_HW_DECODE : RaliDecodeDevice::RALI_SW_DECODE);
```
[the remark this answers is not in the document]
Comment (Indumathi31) [reply utilities/rali/rali_video_unittests/rali_video_unittests.cpp:88]: Done

File: rocAL/rocAL/source/hardware_video_decoder.cpp (hunk 2 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/hardware_video_decoder.cpp:56]: Yes. This function is used to get the hw decoder pixel format. I tried to keep it as a member of HardwareVideoDecoder but it failed. Will fix this and create PR

File: rocAL/rocAL/source/hardware_video_decoder.cpp (hunk 1 above)
Comment (Indumathi31) [reply rocAL/rocAL/source/hardware_video_decoder.cpp:48]: Done

File: rocAL/rocAL/source/hardware_video_decoder.cpp [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+HardWareVideoDecoder::HardWareVideoDecoder(){};
+
+int HardWareVideoDecoder::seek_frame(AVRational avg_frame_rate, AVRational time_base, unsigned frame_number)
+{
+    auto seek_time = av_rescale_q((int64_t)frame_number, av_inv_q(avg_frame_rate), AV_TIME_BASE_Q);
+    int64_t select_frame_pts = av_rescale_q((int64_t)frame_number, av_inv_q(avg_frame_rate), time_base);
+    int ret = av_seek_frame(_fmt_ctx, -1, seek_time, AVSEEK_FLAG_BACKWARD);
+    if (ret < 0)
+    {
+        ERR("HardWareVideoDecoder::seek_frame Error in seeking frame..Unable to seek the given frame in a video");
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/hardware_video_decoder.cpp:37]: remove 2 dots.

**Review by shobana-mcw — COMMENTED** [review]
Did you get_hw_format function?

**Pull request closed without merging**